### PR TITLE
Custom Tensor Registration

### PR DESCRIPTION
**Pull request by bwasti**
This PR introduces a dirt simple registration API to allow custom Tensors.  The API exposed aims to avoid code complexity and compilation times.  It makes an explicit trade-off supporting developer productivity over compute efficiency.

Couple of reasons to have this:
- Tests are easier to write (see #26719)
- It addresses the limit for the 64 possible TensorTypeId's we can have with TensorTypeSet
- Drastically reduces mental overhead for folks working in research wanting to add a Tensor
- Allows entirely decoupled static registration for tensors defined in separate repos (planned usage in pytorch/sparse)

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#27201 Custom Tensor Registration**
* #27159 Use FunctionSchema instead of char* for dispatch

### CODE REVIEW

File: aten/src/ATen/core/CustomTensorImpl.cpp [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+#include <ATen/core/ATenDispatch.h>
+#include <torch/csrc/autograd/variable.h>
+#include <torch/csrc/jit/script/function_schema_parser.h>
+
+namespace at {
+
+c10::intrusive_ptr<TensorImpl> CustomTensorImpl::shallow_copy_and_detach(
+    const c10::VariableVersion& version_counter,
+    bool allow_tensor_metadata_change) const {
+  auto impl = c10::make_intrusive<CustomTensorImpl>(
```
Comment (ezyang) [inline aten/src/ATen/core/CustomTensorImpl.cpp]: nit: just directly return here right?

**Comment by ezyang** [comment]
I agree this is in principle a useful thing.

However, I disagree that now is the right time to add it. The most concrete use case pitched here is for pytorch/sparse, but this is development that *we* are doing, so we should just allocate a tensor type id for it and let it live or die by its merits. I *do* think there is probably something useful with a TensorImpl subclass that handles most of the user boilerplate for you--I think we've been talking about wanting something like this to exist, but I don't want to throw open the gates to end users to be able to extend tensor without talking to us first (because now we are signed up to support this API forever). And it isn't really true out of line allocation: since you're using a string identifier, people have to make sure not to reuse the same string identifier.

File: aten/src/ATen/core/CustomTensorImpl.h [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+#include <ATen/core/TensorBody.h>
+#include <c10/core/TensorImpl.h>
+
+using Impl = void(const c10::FunctionSchema& schema, torch::jit::Stack*);
+
+namespace at {
+
+class CAFFE2_API CustomTensorImpl : public TensorImpl {
+  size_t tag_;
+  std::shared_ptr<void> storage_;
```
Comment (ezyang) [inline aten/src/ATen/core/CustomTensorImpl.h]: Let's not call it a storage, it's very confusing with our native notion of storage (i.e., it doesn't have anything to do with storage!)

File: aten/src/ATen/core/CustomTensorImpl.h [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+#include <unordered_map>
+#include <ATen/core/stack.h>
+#include <ATen/core/TensorBody.h>
+#include <c10/core/TensorImpl.h>
+
+using Impl = void(const c10::FunctionSchema& schema, torch::jit::Stack*);
+
+namespace at {
+
+class CAFFE2_API CustomTensorImpl : public TensorImpl {
```
Comment (ezyang) [inline aten/src/ATen/core/CustomTensorImpl.h]: Besides the tagging, could you explain the other differences between this and OpaqueTensorImpl?

File: aten/src/ATen/native/native_functions.yaml [hunk 4]
```diff
@@ -3404,8 +3404,13 @@
   requires_tensor: True
 
 
+- func: to_custom(Tensor self, str type) -> Tensor
+  variants: method
```
Comment (ezyang) [inline aten/src/ATen/native/native_functions.yaml]: Even considering this API on its merits, it seems kind of strange to me that this has to take a Tensor to create a custom tensor. What exactly am I feeding it here, anyway?

Comment (bwasti) [reply aten/src/ATen/native/native_functions.yaml]: Conceptually, tensors are well represented by specifications of indexing, dtype and value. The underlying storage, device and associated method implementations are irrelevant.  `to_custom` swaps the latter bit without changing the former.

File: aten/src/ATen/core/CustomTensorImpl.h (hunk 2 above)
Comment (bwasti) [reply aten/src/ATen/core/CustomTensorImpl.h]: it's where the user defined tensor's data etc is actually stored, but I agree its confusing with Storage already a reserved name in the codebase

**Comment by bwasti** [comment]
> we should just allocate a tensor type id for it

then we can't work out of source.  I'd like to see the ideas related to sparse etc being ironed out in a non-intrusive way and potentially "upstreamed" to TensorTypeId.  The introduced mechanism lets me move substantially faster and avoid having to RFC everything I'm playing with (see QTensors).  It also forces me to be a customer of my own API and reduces the distinction between "internal" and "external" code.  It's a great way to polish up and refine the API rather than designing it in a void.  We don't need to publicize it yet.

re string tagging -- this is a good point, but can be avoided (and implicitly is) by using the "name" of the tensor as a static symbol.  It would cause compile time conflicts to register the same tensor twice.

File: aten/src/ATen/core/CustomTensorImpl.h (hunk 3 above)
Comment (bwasti) [reply aten/src/ATen/core/CustomTensorImpl.h]: I guess not much besides how it is meant to be used.

File: aten/src/ATen/native/native_functions.yaml (hunk 4 above)
Comment (gchanan) [reply aten/src/ATen/native/native_functions.yaml]: dumb question: why do we need it?  I.e. why can't someone write a per-custom-type extension function for doing this?

**Comment by gchanan** [comment]
> then we can't work out of source

I haven't dug into the details here that much, but I don't think we should be overly-binary about this.  There's a sliding scale of how much internal involvement is necessary, and it's not obvious to me that immediate jumping to one end of the scale is the correct approach.

**Comment by dzhulgakov** [comment]
I echo Ed - for techdebt simplification we should resolve how we handle Backends and such. For me, a good balance might be that all you need in core is extend one enum and then let the rest happen out of tree.

File: aten/src/ATen/core/TensorBody.h [hunk 5]
```diff
@@ -747,6 +747,8 @@ class CAFFE2_API Tensor {
   Tensor & sparse_resize_(IntArrayRef size, int64_t sparse_dim, int64_t dense_dim) const;
   Tensor & sparse_resize_and_clear_(IntArrayRef size, int64_t sparse_dim, int64_t dense_dim) const;
   Tensor sparse_mask(const Tensor & mask) const;
+  Tensor to_custom(std::string type) const;
```
Comment (bwasti) [inline aten/src/ATen/core/TensorBody.h]: better as factory functions + pybind

**Pull request closed without merging**